### PR TITLE
Add support for SPN

### LINKED ISSUE #66: Use Service Principal secrets instead of SAS token
First of all, thanks a ton for the Addon!! Awesome work!

For long term usability, I'd like to use a service principal and client secret to authenticate `azcopy`.

Currently the SAS token generator only supports SAS token validity upto 7 days.

Also, using Azure Active Directory (Service Principal) is recommended over SAS token in general.

### Possible Implementation:

```yaml
# config.yaml
tenant_id: [TENANT-ID]
storage_url: https://[account].blob.core.windows.net/[container]/[path/to/virtual/dir]
principal_id: [PRINCIPAL-ID]
client_secret: [!secret]
delete: false
```

According to `azcopy login` [documentation:](https://learn.microsoft.com/en-us/azure/storage/common/storage-ref-azcopy-login?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json#examples) we can use the cmd below to login with service principal and secret:

```bash
# Export AZCOPY_SPA_CLIENT_SECRET with the secret, then run
azcopy login --service-principal --application-id <your service principal's application ID>
```

### PR DESCRIPTION
**Pull request by jak119**

### CODE REVIEW

File: azcopy/run.sh [hunk 1]
```diff
@@ -11,13 +16,30 @@ fi
 params+=(--output-level essential)
 params+=(--put-md5)
 
+case $auth_type in
+  sas)
+    bashio::log.info "Using SAS token";;
+
+  sp)
+    [[ -z $tenant_id || -z $client_id || -z $client_secret ]] && bashio::exit.nok "tenant_id, client_id and client_secret are required for auth_type=sp"
```
Comment (jak119) [inline azcopy/run.sh:24]: @kforeverisback I'm having trouble getting past line 24 in my testing. I get the log output I added, but line 24 seems to just silently exit.

```text
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
[00:27:24] INFO: Using Service Principal
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped```

Any thoughts?

File: azcopy/run.sh [hunk 2]
```diff
@@ -11,13 +16,31 @@ fi
 params+=(--output-level essential)
 params+=(--put-md5)
 
+case $auth_type in
+  sas)
+    bashio::log.info "Using SAS token. Any Service Principal values (client ID, etc) will be ignored";;
+
+  sp)
+    bashio::log.info "Using Service Principal"
+    [[ -z $tenant_id || -z $client_id || -z $client_secret ]] && bashio::exit.nok "tenant_id, client_id and client_secret are required for auth_type=sp"
```
Comment (jak119) [inline azcopy/run.sh]: @kforeverisback I'm having trouble getting past line 25 in my testing with a SPN. I get the log output I added, but line 25 seems to just silently exit.

```text
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
[00:27:24] INFO: Using Service Principal
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped```

Any thoughts?

Comment (jak119) [reply azcopy/run.sh]: Refactored this logic with bashio functions and all seems to be well (although error handling isn't great).

**Comment by kforeverisback** [comment]
@jak119 I know this is closed, but just wanted to say thanks a ton for finishing this up.
I was out for Thanksgiving, I couldn't check Github's notfication.

**Comment by jak119** [comment]
My pleasure! Hope it works and thanks for the help in getting it implemented!

**Pull request merged**